src/Interpreters/ClusterDiscovery.cpp

```diff
@@ -0,0 +1,276 @@
+#include <algorithm>
+#include <chrono>
+#include <unordered_set>
+
+#include <base/getFQDNOrHostName.h>
+#include <base/logger_useful.h>
+
+#include <Common/DNSResolver.h>
+#include <Common/StringUtils/StringUtils.h>
+#include <Common/ZooKeeper/Types.h>
+#include <Common/setThreadName.h>
+
+#include <Core/ServerUUID.h>
+
+#include <Interpreters/Cluster.h>
+#include <Interpreters/ClusterDiscovery.h>
+#include <Interpreters/Context.h>
+
+namespace DB
+{
+
+namespace
+{
+
+constexpr size_t MAX_QUEUE_SIZE = 16;
+constexpr UInt64 QUEUE_OP_TIMEOUT_MS = 1000;
+
+fs::path getReplicasListPath(const String & zk_root)
+{
+    return fs::path(zk_root + "/replicas");
+}
+
+}
+
+ClusterDiscovery::ClusterDiscovery(
+    const Poco::Util::AbstractConfiguration & config,
+    ContextMutablePtr context_,
+    const String & config_prefix)
+    : context(context_)
+    , node_name(toString(ServerUUID::get()))
+    , server_port(context->getTCPPort())
+    , queue(std::make_shared<UpdateQueue>(MAX_QUEUE_SIZE))
+    , log(&Poco::Logger::get("ClusterDiscovery"))
+{
+    Poco::Util::AbstractConfiguration::Keys config_keys;
+    config.keys(config_prefix, config_keys);
+
+    for (const auto & key : config_keys)
+    {
+        String path = config.getString(config_prefix + "." + key + ".path");
+        trimRight(path, '/');
+        clusters_info.emplace(key, ClusterInfo(key, path));
```

Comment:
Maybe it worth checking that such clusters does not exists in non-autodiscovery mode?

Reply:
Now I placed in under same `remote_servers` section, like https://github.com/vdimir/ClickHouse/blob/106d5486da584c98d38e4ae58e6a00defa126ec1/tests/integration/test_cluster_discovery/config/config.xml so it solves problem